include/irr/asset/material_compiler/IR.h

```diff
@@ -0,0 +1,429 @@
+                    if (prevSource == EPS_TEXTURE)
+                        value.texture.~STextureSource();
+                    value.constant = rhs.value.constant;
+                }
+                else {
+                    if (prevSource == EPS_CONSTANT)
+                        value.constant = type_of_const();
+                    value.texture = rhs.value.texture;
+                }
+
+                return *this;
+            }
+
+            bool operator==(const SParameter<type_of_const>& rhs) const {
+                if (source!=rhs.source)
+                    return false;
+                switch (source)
+                {
+                case EPS_CONSTANT:
+                    return value.constant==value.constant;
+                case EPS_TEXTURE:
+                    return value.texture==value.texture;
+                default: return false;
+                }
+            }
+
+            E_PARAM_SOURCE source;
+            UTextureOrConstant<type_of_const> value;
+        };
+
+        _IRR_STATIC_INLINE_CONSTEXPR size_t MAX_CHILDREN = 16ull;
+        struct children_array_t {
+            INode* array[MAX_CHILDREN] {};
+            size_t count = 0ull;
+
+            inline bool operator!=(const children_array_t& rhs) const
+            {
+                if (count != rhs.count)
+                    return true;
+
+                for (uint32_t i = 0u; i < count; ++i)
+                    if (array[i]!=rhs.array[i])
+                        return true;
+                return false;
+            }
+            inline bool operator==(const children_array_t& rhs) const
+            {
+                return !operator!=(rhs);
+            }
+
+            inline bool find(E_SYMBOL s, size_t* ix = nullptr) const 
+            { 
+                auto found = std::find_if(begin(),end(),[s](const INode* child){return child->symbol==s;});
+                if (found != (array+count))
+                {
+                    if (ix)
+                        ix[0] = found-array;
+                    return true;
+                }
+                return false;
+            }
+
+            operator bool() const { return count!=0ull; }
+
+            INode** begin() { return array; }
+            const INode*const* begin() const { return array; }
+            INode** end() { return array+count; }
+            const INode*const* end() const { return array+count; }
+
+            inline INode*& operator[](size_t i) { assert(i<count); return array[i]; }
+            inline const INode* const& operator[](size_t i) const { assert(i<count); return array[i]; }
+        };
+        template <typename ...Contents>
+        static inline children_array_t createChildrenArray(Contents... children) 
+        { 
+            children_array_t a;
+            const INode* ch[]{ children... };
+            memcpy(a.array, ch, sizeof(ch));
+            a.count = sizeof...(children);
+            return a; 
+        }
```

Comment:
leave a TODO comment...

we could implement the children array in the future as N nodes allocated just after this one (one would only need the child count)... but this would only be possible if the nodes were uniform bytesize.

That way there would be no artificial limit on max children in our IR (backends will still have limits)

Reply:
leave the comment